lib/src/runner/reporter/xunit.dart

```diff
@@ -0,0 +1,353 @@
+    }
+
+    _subscriptions.add(liveTest.onError
+        .listen((error) => _onError(liveTest, error.error, error.stackTrace)));
+
+    _subscriptions.add(liveTest.onPrint.listen((line) {}));
+  }
+
+  /// Return an ordered list of groups that [liveTest] belongs to
+  List _orderedGroupList(List<Group> listOfGroups) {
+    List separatedList = [];
+    if (listOfGroups.length > 1) {
+      separatedList.add(listOfGroups[1].name);
+      for (int i = 2; i < listOfGroups.length; i++) {
+        separatedList.add(listOfGroups[i]
+            .name
+            .replaceAll(listOfGroups[i - 1].name, '')
+            .trim());
+      }
+    }
+    return separatedList;
+  }
+
+  /// A callback called when [liveTest]'s state becomes [state].
+  void _onStateChange(LiveTest liveTest, State state) {
+    if (state.status != Status.complete) {
+      if (state.status == Status.running &&
+          !_testCases.containsKey(liveTest.test)) {
+        String testName = liveTest.individualName;
+        testName = testName.replaceAll('&', '&amp;');
+        testName = testName.replaceAll('<', '&lt;');
+        testName = testName.replaceAll('>', '&gt;');
+        testName = testName.replaceAll('"', '&quot;');
+        testName = testName.replaceAll("'", '&apos;');
+
+        _testCases[liveTest] = new XunitTestResult(
+            testName, liveTest.suite.path, _stopwatch.elapsedMilliseconds);
+        List listOfGroups = _orderedGroupList(liveTest.groups);
+        Map listGroup = _groupStructure;
+
+        listOfGroups.forEach((elem) {
+          if (!listGroup.containsKey(elem)) {
+            listGroup[elem] = {'testResults': new XunitTestSuite()};
+          }
+          listGroup = listGroup[elem];
+        });
+
+        listGroup['testResults'] ??= new XunitTestSuite();
+
+        listGroup['testResults'].add(_testCases[liveTest]);
+      }
+      return;
+    }
+
+    _testCases[liveTest].endTime = _stopwatch.elapsedMilliseconds;
+
+    if (liveTest.test.metadata.skip) {
+      _testCases[liveTest].skipped = true;
+      if (liveTest.test.metadata.skipReason != null) {
+        _testCases[liveTest].skipReason = liveTest.test.metadata.skipReason;
+      }
+    }
+  }
+
+  /// A callback called when [liveTest] throws [error].
+  void _onError(LiveTest liveTest, error, StackTrace stackTrace) {
+    if (liveTest.state.status != Status.complete) return;
+    if (_testCases[liveTest].error.isEmpty) {
+      if (error is TestFailure) {
+        _failureCount++;
+        _testCases[liveTest].error['failure'] = true;
+      } else {
+        _errorCount++;
+        _testCases[liveTest].error['failure'] = false;
+      }
+    }
+    String body = terseChain(stackTrace).toString().trim();
+    body = body.replaceAll('<', '&lt;');
+    body = body.replaceAll('>', '&gt;');
+    String name = error.toString().replaceAll('\n', '');
+    _testCases[liveTest].error['list'] ??= [];
+    _testCases[liveTest].error['list'].add(new XunitFailureResult(name, body));
```

Comment:
There's quite a few places where you're using a map to store various information (error list, whether or not it's a failure, test results). It would be easier to understand if it was done in a typed manner (using a class) instead of an untyped map